docs/linking.md

```diff
@@ -0,0 +1,218 @@
+
+Entry points can also be specified in source via the `[shader()]` attribute also. For example
+
+```
+[shader("compute")]
+void computeMain(uint3 dispatchThreadID : SV_DispatchThreadID)
+{
+    // ... 
+}
+```
+
+Slang libraries are stored as serialized Slang IR. That Slang IR currently maintains no forward or backward compatibility, and a new version of Slang may either produce incompatible IR, or be unable to consume previous versions IR.
+
+By default Slang stores serialized Slang IR in a compressed `lite` format. It can also be stored without any addition compression. These options can be specified via the command line via
+
+```
+-ir-compression lite
+-ir-compression none
+```
+
+They are also available via the API through the `spProcessCommandLineArguments` function. 
+
+Note that Slang can consume and process either `lite` or `none` styles transparently. Also mixing compressed libraries with uncompressed libraries also works. 
+
+Symbols
+-------
+
+Linkage is the process of combining of libraries and source to produce something. The mechanism used with Slang, is that items that are going to be imported or exported are named. As the Slang language supports function overloading and generics, it is not possible to just use a function name, or type name to uniquely identify the item. Thus exported and imported names are 'mangled' to include this additional information. This is not unlike how C++ produces mangled names for symbols for similar reasons.
+
+Symbols within a module are additionally uniquely marked with the name of the module, and thus the module name becomes part of the mangled name. This has implications for includes. For example say we have two modules, and they both #include a header file that contains a function. That function will appear as two separate functions - one defined in each module, because a #include is a textual the function will be defined once in each module and with each module name.
+
+By default Slang will provide linkage symbols for all global symbols - types or functions, including declarations. In the future this is likely to change with declarations *not* being exported by default, as the declaration is made available just to forward define within the current compilation, not as a mechnism to descibe something to be imported. 
+
+To mark a declaration as having it's definition defined elsewhere outside the current module, the `[__extern]` attribute can be used. For example 
+
+```
+[__extern] Thing makeThing(int a, int b);
+``` 
+
+With types the `[__extern]` attribute can also be used. For the moment though a type declaration that is externed still needs to be a type definition. Therefore to declare a type to be imported from elsewhere we actually have to write
```

Comment:
In case it isn't clear, the weird thing with having to write:

```
struct Thing {}
```

instead of:

```
struct Thing;
```

Is a non-issue in the proper approach to linkage where things are defined in different modules. I'm not sweating it in our current system simply because `[__extern]` is not here to stay.